proposals/0000-dictionary-key-and-value-collections.md

```diff
@@ -0,0 +1,189 @@
+}
+```
+
+## Detailed design
+
+* The standard library introduces two new collection types: `DictionaryKeys` and `DictionaryValues`.
+* A `Dictionary`'s `keys` and `values` properties change from `LazyMapCollection` to these new types. 
+* The new collection types are not directly constructable. They are presented only as views into a dictionary.
+
+```swift
+struct Dictionary<Key: Hashable, Value>: ... {
+    var keys: DictionaryKeys<Key, Value> { get }
+    var values: DictionaryValues<Key, Value> { get set }
+    
+    // Remaining declarations
+}
+
+/// A collection view of a dictionary's keys.
+struct DictionaryKeys<Key: Hashable, Value>: Collection {
+    typealias Index = DictionaryIndex<Key, Value>
+    subscript(i: Index) -> Key { get }
+
+    // Other `Collection` requirements
+}
+
+/// A mutable collection view of a dictionary's values.
+struct DictionaryValues<Key: Hashable, Value>: MutableCollection {
+    typealias Index = DictionaryIndex<Key, Value>
+    subscript(i: Index) -> Value { get set }
+
+    // Other `Collection` requirements
+}
+```
+
+A sample implementation of this proposal can be found in [this branch](https://github.com/apple/swift/compare/master...natecook1000:nc-dictionary).
+
+
+## Impact on existing code
+
+The performance improvements of using the new `DictionaryKeys` type and the mutability of the `DictionaryValues` collection are both additive in nature.
+
+Most uses of these properties are transitory in nature. Adopting this proposal should not produce a major impact on existing code. The only impact on existing code exists where a program explicitly specifies the type of a dictionary's `keys` or `values` property. In those cases, the fix would be to change the specified type.
+
+
+## Alternatives considered
+
+
+1. The [Generics Manifesto](https://github.com/apple/swift/blob/master/docs/GenericsManifesto.md) lists nested generics as a goal. This could impact the naming and structure of these new collection types. 
+
+   Instead of `DictionaryKeys<Key, Value>` and `DictionaryValues<Key, Value>`, these types could be `Dictionary<Key, Value>.Keys` and `Dictionary<Key, Value>.Values`. However, because many types in the standard library may be revisited once such a feature is available (indices, iterators, etc.), the current lack of nesting shouldn't prevent consideration of this proposal.
+
+2. Add additional compiler features that manage mutation through existing key-based subscripting without the copy-on-write problems of the current implementation. This could potentially be handled by upcoming changes to copy-on-write semantics and/or inout access. 
+
+3. Provide new APIs for updating dictionary values with a default value, eliminating the double-lookup for a missing key. The approach outlined in this proposal provides a way to remove one kind of double-lookup (mutating a value that exists) but doesn't eliminate all of them (in particular, checking for the existence of a key before adding).
```

Comment:
This is such an important use case that I think it should be elevated to be part of the proposal (in the `dict["one", withDefault: []].append(1)` form would be my personal preference).

For completeness, you might also want to add the other possibility of supplying a default value or closure to calculate one as part of dictionary creation. Personally I think that option is overkill but while we're considering all the options...